### DISCUSSION TITLE
How to use hloc for localization

### DISCUSSION THREAD

**Discussion by kurbster** [zju3dv/LoFTR#101] [Q&A]
How did you use LoFTR to generate the matches for hloc? Hloc expects the output as indices of key point matches. i.e

```matches0 = [k0, k1, -1, .... ,kn]```

This implies the first key point in the first image matches with the _k0th_ key point in the second while the third key point in the first image did not match with any key point in the second so that array value is `-1`.

LoFTR outputs coordinates of the matches and most of the time the number of matches is greater than the number of key points.

How were you able to use the hloc pipeline? What did you have to change?

Thank you any advice is appreciated!

**Comment by zehongs** [comment]
Hi,
I think you can use hloc-toolbox and loftr directly, as inloc is probably providing depth map for each database image.
But for aachen, we round the outputs of LoFTR to integer.

**Reply by kurbster** [reply]
I am currently working with the aachen dataset. 

From the dictionary returned by LoFTR I am using the `mkpts0_f` and `mkpts1_f` as the matches found. I am getting an error because those values are 2-dimensional `[[x0 y0], [x1 y1]...]` whereas hloc is expecting a 1-dimensional answer `[idx0, idx1, ...]`

Am I using the wrong output from LoFTR?

**Reply by zehongs** [reply]
I think you need to convert the LoFTR output to hloc's format?

**Reply by kurbster** [reply]
How did you do it when you published the results? Do you have any advice on how to convert into hloc format?

**Reply by zehongs** [reply]
This issue is related to the topic.
https://github.com/zju3dv/LoFTR/issues/9

**Reply by kurbster** [reply]
Ok, I will take a look at that code and try it out. Just so I understand, for the DB images you need to use round, but for the query images you can leave the floating point values. Also, you don't need to run feature extraction phase before hand because LoFTR is a detector-free matcher. 

Thank you for your help and the quick responses!!!
